include/envoy/network/io_handle.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include <memory>
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * IoHandle
+ */
+class IoHandle {
+public:
+  IoHandle(int fd=-1):fd_(fd) { }
```

Comment:
I'd prefer we keep this a pure interface, and have an IoHandleFdImpl (please come up with a better name! =P) so the constructor isn't part of the interface from the get go.

Also please add a TODO to remove fd() function and friends since we know we want to get rid of it, but folks reading the PR may not know.

Reply:
+1. In general, I would rather we just get rid of all of the implicit casts, assignment, equality, etc. operators. It will make this change a bit more verbose but much easier to search the code for what is left to fix as we gradually remove `fd()` access. This is basically a stronger +1 to @alyssawilk comment below.

Reply:
Got it- please see comment below.